src/OrbitBase/SimpleExecutorTest.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gtest/gtest.h>
+
+#include "OrbitBase/SimpleExecutor.h"
+
+namespace orbit_base {
+
+TEST(SimpleExecutor, ScheduledTaskShouldBeCalledSimpleWithVoid) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  auto future = executor->Schedule([&called]() { called = true; });
+  executor->ExecuteScheduledTasks();
+  EXPECT_TRUE(called);
+  EXPECT_TRUE(future.IsFinished());
+}
+
+TEST(SimpleExecutor, ScheduledTaskShouldBeCalledSimpleWithInt) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  auto future = executor->Schedule([&called]() {
+    called = true;
+    return 42;
+  });
+  executor->ExecuteScheduledTasks();
+  EXPECT_TRUE(called);
+  ASSERT_TRUE(future.IsFinished());
+  EXPECT_EQ(future.Get(), 42);
+}
+
+TEST(SimpleExecutor, ChainedTaskedShouldBeCalledSimple) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  Promise<void> promise{};
+  auto future = promise.GetFuture();
+  auto chained_future = executor->ScheduleAfter(future, [&called]() { called = true; });
+  EXPECT_FALSE(called);
+  EXPECT_FALSE(chained_future.IsFinished());
+  promise.MarkFinished();
+  executor->ExecuteScheduledTasks();
+  EXPECT_TRUE(called);
+  EXPECT_TRUE(chained_future.IsFinished());
+}
+
+TEST(SimpleExecutor, ScheduleAfterIfSuccessShortCircuitOnErrorVoid) {
+  auto executor = SimpleExecutor::Create();
+  bool called = false;
+  Promise<ErrorMessageOr<void>> promise{};
+  auto future = promise.GetFuture();
+  auto chained_future = executor->ScheduleAfterIfSuccess(future, [&called]() { called = true; });
+  EXPECT_FALSE(called);
+  EXPECT_FALSE(chained_future.IsFinished());
+
+  constexpr const char* const kErrorMessage{"Error"};
+  promise.SetResult(ErrorMessage{kErrorMessage});
+  executor->ExecuteScheduledTasks();
```

Comment:
Is `chained_future` already finished between these 2 lines?

Reply:
No. `promise.SetResult` schedules the continuation, and `ExecuteScheduledTasks` executes the continuation and completes the future. Does that make sense?

Reply:
I looked at the code for a while and then decided to just test it. 
The following "is true" (which I tried to ask earlier)
```
promise.SetResult(ErrorMessage{kErrorMessage});
EXPECT_TRUE(chained_future.IsFinished());
executor->ExecuteScheduledTasks();
```

I looked more at the code and it makes sense. This is the "Short Circuit Pathway" in `Executor::ScheduleAfterIfSuccess`. On `promise.SetResult` the continuations of the corresponding future are called and with `ScheduleAfterIfSuccess` this continuation "short circuits" and effectively also calls "`chained_promise.SetResult`" (promise of `chained_future`). In this case another action is scheduled to erase the old functor (`[&called]() { called = true; }`), which was never executed. 

This is all complicated stuff. But I think it all checks out in the end. Also maybe you misunderstood my initial question. 

Just for completion, this is how I augmented the test to understand it:
```
...
  auto future = promise.GetFuture();
  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 0);
  auto chained_future = executor->ScheduleAfterIfSuccess(future, [&called]() { called = true; });
  EXPECT_FALSE(called);
  EXPECT_FALSE(chained_future.IsFinished());

  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 1);
  constexpr const char* const kErrorMessage{"Error"};
  promise.SetResult(ErrorMessage{kErrorMessage});
  EXPECT_TRUE(chained_future.IsFinished());
  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 1);
  executor->ExecuteScheduledTasks();
  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 0);
  EXPECT_FALSE(called);
...
```

Reply:
Ahh, yes. You're right. Sorry. I didn't get that this was about the error case.

Reply:
I quickly thought about it and I'm not entirely sure if that behaviour is an implementation detail or not. Nevertheless I will add the check! On first thought it makes sense to me, we can still remove it in the future!

Reply:
Done.